### PR TITLE
Update the library view to include downloading

### LINKED ISSUE #182: Batch download [Unread/All] from Library
Would love to request this feature if possible. Designed this UI for easy understanding.

Download
Download_pop

### PR DESCRIPTION
**Pull request by zabroseric**
Resolves #182

### CODE REVIEW

**Comment by zabroseric** [comment]
Example of UI
image

Note: I'm a React / NodeJS dev, so feel free to request corrections to better follow project/language standards.

File: iOS/Scenes/Library/LibraryViewController.swift [hunk 1]
```diff
@@ -944,6 +944,25 @@ extension LibraryViewController {
...
+            let chapters = (try? await source?.getChapterList(manga: manga)) ?? []
+            DownloadManager.shared.download(chapters: chapters, manga: manga)
+        }
+    }
+
+    func downloadUnreadAction(manga: Manga) {
+        Task {
+            let source = SourceManager.shared.source(for: manga.sourceId)
+            let readingHistory = await CoreDataManager.shared.getReadingHistory(sourceId: manga.sourceId, mangaId: manga.id)
+            let chapters = ((try? await source?.getChapterList(manga: manga)) ?? []).filter {
```
Comment (Skittyblock) [inline iOS/Scenes/Library/LibraryViewController.swift:959]: probably best to use `CoreDataManager.shared.getChapters` instead https://github.com/Aidoku/Aidoku/blob/d19878092bfe2168d95e7a234b743d55d320580e/Shared/Managers/CoreData/CoreDataManager%2BChapter.swift#L51

**Review by Skittyblock — CHANGES_REQUESTED** [review]
might as well set the two download functions as `async` rather than spawning a task inside them and then instead spawn a task when you call them

**Comment by zabroseric** [comment]
Awesome, thanks @Skittyblock for the info, especially considering it would just be quicker for
you to resolve it yourself. I've fixed what you mentioned, and it makes sense. It feels like the download functionality (all/unread) should be moved out of the controller, and into something like the DownloadManager. Thoughts?

File: iOS/Scenes/Library/LibraryViewController.swift [hunk 2]
```diff
@@ -944,6 +944,19 @@ extension LibraryViewController {
...
 
+    func downloadAll(manga: Manga) async {
+        let chapters = await CoreDataManager.shared.getChapters(sourceId: manga.sourceId, mangaId: manga.id)
+        DownloadManager.shared.download(chapters: chapters, manga: manga)
+    }
+
+    func downloadUnread(manga: Manga) async {
+        let readingHistory = await CoreDataManager.shared.getReadingHistory(sourceId: manga.sourceId, mangaId: manga.id)
+        let chapters = await CoreDataManager.shared.getChapters(sourceId: manga.sourceId, mangaId: manga.id).filter {
+            readingHistory[$0.id] == nil || readingHistory[$0.id]?.page != -1
```
Comment (zabroseric) [inline iOS/Scenes/Library/LibraryViewController.swift:955]: Ensure partially, but unread chapters are downloaded.

**Comment by Skittyblock** [comment]
yeah might be good to add the functions to `DownloadManager` instead since I might add an option in the manga menu for downloading (like unread and ranges, etc.)

**Comment by zabroseric** [comment]
Sweet, that change has been made. Yeah, I was thinking the same and played around with the idea below:

image

**Comment by Skittyblock** [comment]
I would probably just do a child menu in the ellipsis menu rather than a popover but yeah

**Review by Skittyblock — APPROVED** [review]

**Comment by zabroseric** [comment]
Sweet, your project, your call. :)

**Pull request merged**